src/object/srv_obj.c

```diff
@@ -1053,6 +1053,46 @@ obj_fetch_create_maps(crt_rpc_t *rpc, struct bio_desc *biod)
 	return 0;
 }
 
+static int
+obj_fetch_shadow(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
+		uint64_t cond_flags, daos_key_t *dkey, unsigned int iod_nr,
+		daos_iod_t *iods, uint32_t tgt_idx,
+		struct daos_recx_ep_list **pshadows)
+{
+	struct daos_oclass_attr		*oca;
+	daos_handle_t			 ioh = DAOS_HDL_INVAL;
+	int				 rc;
+
+	obj_iod_idx_vos2parity(iod_nr, iods);
+	oca = daos_oclass_attr_find(oid.id_pub);
+	if (oca == NULL || !DAOS_OC_IS_EC(oca)) {
+		rc = -DER_INVAL;
+		D_ERROR(DF_UOID" oca not found or not EC obj: "DF_RC"\n",
+			DP_UOID(oid), DP_RC(rc));
+		goto out;
+	}
+
+	rc = vos_fetch_begin(coh, oid, epoch, cond_flags, dkey, iod_nr, iods,
+			     VOS_FETCH_RECX_LIST, NULL, &ioh, NULL);
+	if (rc) {
+		D_ERROR(DF_UOID" Fetch begin failed: "DF_RC"\n",
+			DP_UOID(oid), DP_RC(rc));
+		goto out;
+	}
+
+	*pshadows = vos_ioh2recx_list(ioh);
+	vos_fetch_end(ioh, 0);
+
+out:
+	obj_iod_idx_parity2vos(iod_nr, iods);
+	if (rc == 0) {
+		obj_iod_idx_vos2daos(iod_nr, iods, tgt_idx, oca);
```

Comment:
we might create another wrapper function to convert both in one call, instead of iterate iod and recx for twice

Reply:
ok, the L1087 is to revert L1066. I'll see and can change it in PR2886 later to avoid conflict.